File: backend/shared/uidl/src/main/java/io/mateu/uidl/data/LayoutDelta.java
package io.mateu.uidl.data;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * What the visual editor changed about a screen's layout, expressed as a DELTA over the inferred
 * one rather than as a snapshot of the result.
 *
 * <p><b>Why this exists.</b> Today the editor writes a full {@code layout:} tree, and an explicit
 * layout takes a screen out of the inference regime for good — <i>explicit always wins</i>, by
 * design. The consequence is quiet and permanent: the moment someone drags one field, that screen
 * stops re-deriving. Add a field to the record and it does not appear; rename one and the layout
 * points at a ghost. That is the divergence that actually matters now — not stale data, which the
 * bundle already solved, but a layout that stopped following its model.
 *
 * <p><b>The shape of the fix.</b> A delta records the DECISIONS a human made — this field first,
 * that one hidden, this one wider — anchored to <b>stable field ids</b> rather than to positions in
 * a tree. Inference still runs on every request; the delta is re-applied on top. A field the model
 * grows is simply one the delta says nothing about, so it appears in its inferred place. A field
 * the model loses is a delta entry that no longer matches anything, and is ignored rather than
 * breaking the page.
 *
 * <p><b>What it deliberately cannot express.</b> Arbitrary restructuring. A delta says "these
 * fields, in this order, with these tweaks" — not "wrap these two in a card inside a tab".
 * Anchoring to ids is exactly what makes it survive model change, and a delta that could rebuild
 * the tree freely would be a snapshot wearing a different name.
 */
public record LayoutDelta(
    List<String> order, List<String> hidden, Map<String, FieldOverride> overrides) {

  /** Per-field tweaks a human made. Every member is optional; null means "leave as inferred". */
  public record FieldOverride(String label, Integer colspan, String section) {}

  public LayoutDelta {
    order = order == null ? List.of() : List.copyOf(order);
    hidden = hidden == null ? List.of() : List.copyOf(hidden);
    overrides = overrides == null ? Map.of() : Map.copyOf(overrides);
  }

  public static LayoutDelta empty() {
    return new LayoutDelta(List.of(), List.of(), Map.of());
  }

  public boolean isEmpty() {
    return order.isEmpty() && hidden.isEmpty() && overrides.isEmpty();
  }

  /**
   * The field ids to render, in order, given what inference produced.
   *
   * <p>The rule that makes this a delta and not a snapshot: <b>fields the delta does not mention
   * keep their inferred position</b>. Listed ones come first in the order the human chose; the rest
   * follow in the order inference gave them. So a field added to the model later appears — which is
   * precisely what a stored snapshot cannot do.
   */
  public List<String> applyTo(List<String> inferred) {
    var result = new ArrayList<String>();
    for (var id : order) {
      // A delta entry for a field the model no longer has is stale, not fatal: ignore it.
      if (inferred.contains(id) && !hidden.contains(id) && !result.contains(id)) {
        result.add(id);
      }
    }
    for (var id : inferred) {
      if (!hidden.contains(id) && !result.contains(id)) {
        result.add(id);
      }
    }
    return result;
  }

  /** The override for a field, or an empty one — so callers never branch on null. */
  public FieldOverride overrideFor(String fieldId) {
    return overrides.getOrDefault(fieldId, new FieldOverride(null, null, null));
  }

  /**
   * The delta that turns {@code inferred} into {@code desired} — what the editor should SAVE after
   * a human rearranges a screen, instead of the whole tree.
   *
   * <p>It records only what differs: a screen dragged into exactly its inferred order produces an
   * empty delta, and therefore keeps re-deriving. That is the property to preserve — using the
   * editor at all should not, by itself, freeze a screen.
   */
  public static LayoutDelta between(List<String> inferred, List<String> desired) {
    var hidden = new ArrayList<String>();
    for (var id : inferred) {
      if (!desired.contains(id)) {
        hidden.add(id);
      }
    }
    var visibleInferred = new ArrayList<>(inferred);
    visibleInferred.removeAll(hidden);
    // Only record an order when it actually differs from what inference already gives.
    var order = visibleInferred.equals(desired) ? List.<String>of() : List.copyOf(desired);
    return new LayoutDelta(order, hidden, new LinkedHashMap<>());
  }
}

File: frontend/web/monorepo/apps/visual-editor/src/model/layoutDelta.ts
/**
 * The editor's half of `layoutDelta:` — the TypeScript mirror of Java's
 * `io.mateu.uidl.data.LayoutDelta`.
 *
 * The whole point of the shape: a `layout:` is a **snapshot**, and an explicit layout takes a screen
 * out of the inference regime for good. The moment someone drags one field, that screen stops
 * re-deriving — add a field to the model afterwards and it never appears. A delta records the
 * DECISIONS instead, anchored to stable field ids, so inference keeps running and the decisions are
 * re-applied on top.
 *
 * These functions must agree with the server exactly, or the editor would show one arrangement and
 * the running page another. `applyDelta` and `deltaBetween` are ports of `applyTo` and `between`,
 * assertion for assertion.
 */

/** Per-field tweaks a human made. Every member is optional; absent means "leave as inferred". */
export interface FieldOverride {
    label?: string
    colspan?: number
    section?: string
}

export interface LayoutDelta {
    order: string[]
    hidden: string[]
    overrides: Record<string, FieldOverride>
}

/** A field as inference produced it — what the server's `__contract__` reports. */
export interface InferredField {
    id: string
    label?: string
    dataType?: string
    stereotype?: string
    required?: boolean
    readOnly?: boolean
}

export function emptyDelta(): LayoutDelta {
    return { order: [], hidden: [], overrides: {} }
}

export function isEmptyDelta(delta: LayoutDelta): boolean {
    return delta.order.length === 0
        && delta.hidden.length === 0
        && Object.keys(delta.overrides).length === 0
}

/**
 * The field ids to render, in order, given what inference produced.
 *
 * The rule that makes this a delta and not a snapshot: **fields the delta does not mention keep
 * their inferred position**. So a field added to the model later appears — precisely what a stored
 * snapshot cannot do.
 */
export function applyDelta(inferred: string[], delta: LayoutDelta): string[] {
    const result: string[] = []
    for (const id of delta.order) {
        // A delta entry for a field the model no longer has is stale, not fatal: ignore it.
        if (inferred.includes(id) && !delta.hidden.includes(id) && !result.includes(id)) {
            result.push(id)
        }
    }
    for (const id of inferred) {
        if (!delta.hidden.includes(id) && !result.includes(id)) result.push(id)
    }
    return result
}

/**
 * The delta that turns `inferred` into `desired` — what the editor saves after a human rearranges a
 * screen, instead of the whole tree.
 *
 * It records only what differs: a screen dragged into exactly its inferred order produces an empty
 * delta and therefore keeps re-deriving. Using the editor at all must not, by itself, freeze a
 * screen.
 */
export function deltaBetween(
    inferred: string[],
    desired: string[],
    overrides: Record<string, FieldOverride> = {},
): LayoutDelta {
    const hidden = inferred.filter((id) => !desired.includes(id))
    const visibleInferred = inferred.filter((id) => !hidden.includes(id))
    const sameOrder = visibleInferred.length === desired.length
        && visibleInferred.every((id, i) => id === desired[i])
    return {
        order: sameOrder ? [] : [...desired],
        hidden,
        overrides: prune(overrides),
    }
}

/** Drop overrides that say nothing, so an untouched field never lands in the file. */
function prune(overrides: Record<string, FieldOverride>): Record<string, FieldOverride> {
    const kept: Record<string, FieldOverride> = {}
    for (const [id, o] of Object.entries(overrides)) {
        const trimmed: FieldOverride = {}
        if (o.label !== undefined && o.label !== '') trimmed.label = o.label
        if (o.colspan !== undefined && o.colspan !== 0) trimmed.colspan = o.colspan
        if (o.section !== undefined && o.section !== '') trimmed.section = o.section
        if (Object.keys(trimmed).length) kept[id] = trimmed
    }
    return kept
}

/** Read a `layoutDelta:` node as parsed from YAML, tolerating a partial or malformed one. */
export function readDelta(node: unknown): LayoutDelta {
    if (!node || typeof node !== 'object') return emptyDelta()
    const raw = node as Record<string, unknown>
    const overrides: Record<string, FieldOverride> = {}
    if (raw.overrides && typeof raw.overrides === 'object') {
        for (const [id, value] of Object.entries(raw.overrides as Record<string, unknown>)) {
            if (value && typeof value === 'object') {
                const o = value as Record<string, unknown>
                overrides[id] = {
                    label: typeof o.label === 'string' ? o.label : undefined,
                    colspan: typeof o.colspan === 'number' ? o.colspan : undefined,
                    section: typeof o.section === 'string' ? o.section : undefined,
                }
            }
        }
    }
    return {
        order: Array.isArray(raw.order) ? raw.order.filter((s) => typeof s === 'string') : [],
        hidden: Array.isArray(raw.hidden) ? raw.hidden.filter((s) => typeof s === 'string') : [],
        overrides: prune(overrides),
    }
}

/** The delta as a plain object for YAML, omitting the parts that carry no decision. */
export function writeDelta(delta: LayoutDelta): Record<string, unknown> {
    const out: Record<string, unknown> = {}
    if (delta.order.length) out.order = delta.order
    if (delta.hidden.length) out.hidden = delta.hidden
    if (Object.keys(delta.overrides).length) out.overrides = delta.overrides
    return out
}
